File: Python3/0378_kth_smallest_element_in_a_sorted_matrix.py
# https://leetcode-cn.com/problems/kth-smallest-element-in-a-sorted-matrix/

class Solution:
    def kthSmallest(self, matrix, k):
        """
        Inputs:
            matrix: list[list[int]]
            k: int
        Outputs:
            res: int
        """

        n = len(matrix)
        nums = [(matrix[i][0], i, 0) for i in range(n)]
        # Convert nums to a heap.
        heapq.heapify(nums)
        for i in range(k-1):
            num, r, c = heapq.heappop(nums)
            if c < n-1:
                heapq.heappush(nums, (matrix[r][c+1], r, c+1))

        res, r, c = heapq.heappop(nums)
        return res

File: Cpp/0378_kth_smallest_element_in_a_sorted_matrix.cpp
// https://leetcode.cn/problems/kth-smallest-element-in-a-sorted-matrix/

class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        auto cmp = [&](const pair<int,int>& p1, const pair<int,int>& p2) {
            return matrix[p1.first][p1.second] > matrix[p2.first][p2.second];
        };

        int rows = matrix.size(), cols = matrix[0].size();
        priority_queue<pair<int,int>, vector<pair<int,int>>, decltype(cmp)> pq(cmp);
        for (int i = 0; i < rows; ++i) {
            pq.push(make_pair(i, 0));
        }

        for (int step = 1; step < k; ++step) {
            pair<int,int> idxs = pq.top();
            pq.pop();
            if (idxs.second + 1 < cols) {
                pq.push(make_pair(idxs.first, idxs.second+1));
            }
        }
        int res = matrix[pq.top().first][pq.top().second];
        return res;
    }
};
